### ISSUE TITLE
Flash.vim style search in a document.

### ISSUE THREAD

**Issue by stephen776** [zed-industries/zed#14801]
### Check for existing issues

- [X] Completed

### Describe the feature

[Flash](https://github.com/folke/flash.nvim) lets you search and jump to Text inside a file by appending jump labels as you type search text. 

it is an incredibly fast way to jump the cursor to specific places and would be an excellent addition to vim mode in Zed.

**Comment by notpeter** [comment]
Hi @stephen776, interesting.  So as I understand this, when searching in vim mode it dynamically creates "labels" which are appended inline following matches to allow you  to quickly jump to that match if you type those letters rather than typing the vim pattern.  These labels appear to be generated dynamically to not conflict with any characters that could following your current pattern. Additionally, it appears to support having the same labels for block starts/end (I assume using tree-sitter)?

What happens if you hit a typo that matches a label? Would this only be possible in vim mode or do you thin kit would be possible to support this in non-vim modes?

Can you say a little bit more about how you use this feature and why you find it so powerful?

**Comment by zebede22** [comment]
I'm using flash.nvim quite actively and supporting something similar on Zed would be amazing.

1. Search integration is great, it allows to jump anywhere on the screen very fast!
![image](https://github.com/user-attachments/assets/bfeff020-b2a8-407d-99bc-3c215a133285)

2. I found treesitter way to select objects very fast as well, with a key press you can select objects
![image](https://github.com/user-attachments/assets/593a4462-11c7-492b-8c7e-5c3f16f7a418)

**Comment by aegatlin** [comment]
Howdy! 😄 

I wanted to share some further "prior art" on this style of (n)vim navigation: 
- vim-easymotion https://github.com/easymotion/vim-easymotion
- VSCodeVim configuration setting to emulate easymotion https://github.com/VSCodeVim/Vim?tab=readme-ov-file#vim-easymotion

Would love to see it in Zed ❤️ and thank you for this awesome editor 🙏

**Comment by machin3io** [comment]
As awesome as zed already is, integrating something like this, even as an extension, would lift it to another level :) Hope it happens.

**Comment by djmango** [comment]
Is this possible to add via an extension? Would prefer core ideally but I'd be happy to look into writing a plugin

**Comment by Dev-cmyser** [comment]
also wait this feature

**Comment by Dev-cmyser** [comment]
> Is this possible to add via an extension? Would prefer core ideally but I'd be happy to look into writing a plugin

yeah, maybe try to do this ?
I don't have ex in rust but i think you , me and chatgpt can create this

**Comment by Dev-cmyser** [comment]
for example - we can create 1 feature - "jump in all creen"

**Comment by djmango** [comment]
As it seems, extensions don't support this kind of functionality

**Comment by Dev-cmyser** [comment]
> As it seems, extensions don't support this kind of functionality

why ? i don't undeerstand

**Comment by djmango** [comment]
Not sure but see the other related issues, they denied a PR for this functionality too

----

Sulaiman Khan Ghori ( https://skg.gg/ ) | X ( https://x.com/sulaimanghori )

----

**Comment by j4ger** [comment]
As far as I'm concerned it isn't possible for extensions to implement this -- at least we need to be able to read user input in some way (for the input of search pattern and follow-up label) and update text on the screen (this is somewhat doable via lsp capabilities but the implementation would be awkward). In the end we still need the devs to come up with other solutions -- either to provide more apis for the extensions (which may benefit other kinds of extensions as well, I personally would love some sort of fzf-lua equivalent) or a native implementation for this feature.

**Comment by nnra6864** [comment]
+1, seems pretty useful.

**Comment by Marrieto** [comment]
I really want to switch to using Zed as my daily, but not having this functionality just makes me so much slower. Until then I'll stick to Rider

**Comment by NightSling** [comment]
Would love this!

**Comment by itpropro** [comment]
Any updates on supporting flash like movements? I was surprised coming from VSCode that something like flash isn’t either built-in or could be added by extension in Zed.

**Comment by joaqo** [comment]
Same, this would be absolutely amazing, 137 thumbs up on the original post and counting!

**Comment by hsyed** [comment]
Big please 🙏❤️!

**Comment by ghalestrilo** [comment]
Just pitching in - what's missing for this to make its way to zed? Can we help at all?

This is easily the biggest feature I miss from other editors. Let's make it happen.

**Comment by vitornesello** [comment]
YES!

**Comment by 0xrusowsky** [comment]
are there any plans to support these features? would be really useful to improve navigation

**Comment by kylejbrk** [comment]
Doesnt helix also support this type of searching put of the box? Seems like it would also be needed to support the helix bindings.

**Comment by rawnly** [comment]
instead of spamming comments of "yes / +1" just react to the issue please! It's enough for the maintainers to understand that people want it

**Comment by walker-tx** [comment]
Howdy! Commenting to add my voice to the rumble!

Really excited to see how much progress has been made on Zed's vim mode! `flash.nvim` is something I lean heavily on in nvim, and it would be nice to see this in Zed as well!

**Comment by didi0987** [comment]
vscode has the same feature through  extension called "cvim". hope Zed can has this feature soon as many people lean on it.

**Comment by margaretjoanmiller** [comment]
This hinging on better extension support really highlights how much a better extension API is needed

**Comment by paul-manole** [comment]
I would suggest [`leap`](https://github.com/ggandor/leap.nvim) instead of [`flash`](https://github.com/folke/flash.nvim). It's:
- more popular (4.9K start vs 3.5k)
- it improves on `vim-sneak` (so you can cross `vim-sneak` off the list of alternative implementations, IMO)
- more ergonomic (allows editing query, eg. backspace lets you adjust/fix queries, and is easy to trigger, with one key)
- it's written in Fennel (LISP for Lua, might be harder to port, but technically AI should be able to parse it more easily)
- it's the default for LazyVim (https://www.lazyvim.org/extras/editor/leap#leapnvim)
- replaces `s` and `S` in vim mode, which aren't all that useful or used, so it wouldn't need to take any other key binds.

**Comment by LogicKahanHai** [comment]
@paul-manole 

> more ergonomic (allows editing query, eg. backspace lets you adjust/fix queries, and is easy to trigger, with one key)

I think flash also gives you the same feature. You can press backspace to edit your query. (Or, I may have misunderstood what you meant)

> t's the default for LazyVim (https://www.lazyvim.org/extras/editor/leap#leapnvim)

No it's not, it is added as an "extra" plugin the default is flash.nvim (https://www.lazyvim.org/plugins/editor#flashnvim) flash is added in the core plugin list.

**Comment by brokenthorn** [comment]
> No it's not, it is added as an "extra" plugin the default is flash.nvim (https://www.lazyvim.org/plugins/editor#flashnvim) flash is added in the core plugin list.

@LogicKahanHai wow, I feel so ashamed. You're right. I misread... Thanks for clarifying!

What I wanted to say is that I like defaults LazyVim provides. The way the "flash/leap" functionality works out of the box feels just perfect to me.

Not that it matters to anyone. But yeah I would also love to have this functionality in Zed. It's probably the only thing I miss, so far. I would make the switch today if this was available in Zed. I hate having to use the classic navigations in VIM and type relative numbers with everything...

**Comment by LogicKahanHai** [comment]
Yes exactly! I really want to shift to Zed, the only thing I am waiting for is this. Flash is so deeply embedded in my daily workflow that I cannot work without it.

**Comment by komarekzm** [comment]
Big please too! Are there any alternatives to jump so quickly in zed?

**Comment by acardonna** [comment]
As others mentioned, this is the missing piece for Zed's Vim integration. And it's also the only reason I haven't switched yet to Zed.

**Comment by jacobedawson** [comment]
Just started trying out Zed and immediately ran into a lack of Flash-like search in Vim mode, currently the only thing that is stopping me from making the switch from vscode. In vscode this is solved by using an extension which is a port of flash.vim, this enables flash-like searching with s and S.

**Comment by arcqiufeng** [comment]
This function is absolutely necessary. I love flash.nivm.

**Comment by brokenthorn** [comment]
Flash.nvim also has wonderful TreeSitter integration. For example, as setup by default in LazyNvim, pressing S (instead of s) triggers TreeSitter-based queries which allow you select nodes and easily visualize and change/delete/append full nodes. You have to try it to understand. It's simply better than all the other similar plugins.

**Comment by tebben** [comment]
Too bad the extension system doesn't allow us to create something like flash.nvim at the moment, hopefully in the near future. Because flash is so embedded in my flow I [added](https://github.com/tebben/zed/tree/feature/jump) something for myself to zed with the help of my friend Claude 4.5. Happily using my local zed build with jump functionality for the time being :)

https://github.com/user-attachments/assets/519d561e-d88b-4a14-a4b0-14cb8eead65f

**Comment by Dima-369** [comment]
@tebben that looks very interesting. I'll definitely check it out and maybe also integrate into my own fork. Thanks!

**Edit: I tested it and I love it ❤️**

Maybe backspace could be handled to delete a char and specifying the keys to use for the hints in the settings (trivial to change in source code), but overall, a great feature!

**Comment by pistvakt005** [comment]
I can only agree with all of the above. Would love to see flash implemented, or possibility to build via extentions or similar.

**Comment by tannerkovach** [comment]
Any update on this?

**Comment by wuchunpeng777** [comment]
I am very happy to share with you the similar functions and features of flash.nvim that I have implemented in Zed Editor

1. Quick jump, support search for multiple characters similar to flash.nvim (not mandatory to search for 2 characters)
2. Based on the number of matching results, the tag will automatically adjust whether it is one character jump or multiple character jumps
3. Support searching for split-screen buffer
4. I did some color adaptation of light and dark themes
5. Characters entered in the tag will be turned down to improve readability

Setting keymap in your keymap.json like this:
```
{
    "context": "vim_mode == normal",
    "bindings": {
      "f": ["vim::Flash", { "backwards": false }]
    }
}
```

Here is my repository [https://github.com/wuchunpeng777/zed.git](url)

https://github.com/user-attachments/assets/8df37c02-0861-463b-8779-297a1759b1cd

**Comment by wuchunpeng777** [comment]
> > Setting keymap in your keymap.json like this:在 keymap.json 中设置键位映射如下：
> > ```
> > {
> >     "context": "vim_mode == normal",
> >     "bindings": {
> >       "f": ["vim::Flash", { "backwards": false }]
> >     }
> > }
> > ```
> 
> [@wuchunpeng777](https://github.com/wuchunpeng777): Good job! And thank you! I hope it gets merged, after it gets polished a bit, if needed.：干得好！谢谢你！我希望它能合并，如果需要的话，先稍微打磨一下。
> 
> By the way, I just wanted to say that most nvim distros bind flash.nvim to `s`, which is the substitution key by default, and is a lot less used/useful than `f`. In fact, I use `f` a lot to jump to the next match (or previous, with capital `F`), mostly for the results of the search function (the `/` key in normal mode).顺便说一句，大多数 nvim 发行版会把 flash.nvim 绑定到 `s` 上，s 是默认的替换键，而且比 `f` 用得少很多。事实上，我经常用 `f` 跳转到下一场比赛（或上一场，大写 `F），主要`是为了搜索功能的结果（普通模式下的 `/` 键）。
> 
> Here is [LazyVim's default setup for flash.nvim](https://www.lazyvim.org/plugins/editor#flashnvim) for the keys reference:这是 [LazyVim 对 flash.nvim 的默认设置 ](https://www.lazyvim.org/plugins/editor#flashnvim)，供密钥参考：
> 
> {
>   "folke/flash.nvim",
>   event = "VeryLazy",
>   vscode = true,
>   ---@type Flash.Config
>   opts = {},
>   -- stylua: ignore
>   keys = {
>     { "s", mode = { "n", "x", "o" }, function() require("flash").jump() end, desc = "Flash" },
>     { "S", mode = { "n", "o", "x" }, function() require("flash").treesitter() end, desc = "Flash Treesitter" },
>     { "r", mode = "o", function() require("flash").remote() end, desc = "Remote Flash" },
>     { "R", mode = { "o", "x" }, function() require("flash").treesitter_search() end, desc = "Treesitter Search" },
>     { "<c-s>", mode = { "c" }, function() require("flash").toggle() end, desc = "Toggle Flash Search" },
>     -- Simulate nvim-treesitter incremental selection
>     { "<c-space>", mode = { "n", "o", "x" },
>       function()
>         require("flash").treesitter({
>           actions = {
>             ["<c-space>"] = "next",
>             ["<BS>"] = "prev"
>           }
>         }) 
>       end, desc = "Treesitter Incremental Selection" },
>   },
> }

Yes, the 'f' key is my personal preferred key binding. I will refer to flash.nvim to set a suitable default key binding and make it available in the keymap for custom configuration. I think there's a reason why flash.nvim binds the 's' key.

I'm currently adding the functionality to jump to the default match. Once this feature is complete, I'll review the code for create a pull request. Additionally, I'd like to gather more feedback on this feature to make it more robust and perfect.

**Comment by szaboboldizsarmark** [comment]
I think helix's jump style is a much better implementation, I think it would be worth considering that instead of flash, or even supporting both, given we have a helix mode.

**Comment by ccidral** [comment]
> I think helix's jump style is a much better implementation, I think it would be worth considering that instead of flash, or even supporting both, given we have a helix mode.

It's better to leave it open to extensions to implement it instead of bolting on a specific strategy.

**Comment by aizigao** [comment]
ok, the new helix amp jump feature also can use in vim mode, in keymap.json
```json
 {
    "context": "(VimControl && !menu)",
    "bindings": {
      "s": "vim::HelixJumpToWord",
    },
  },
```

<img width="1812" height="1342" alt="Image" src="https://github.com/user-attachments/assets/71f0ae60-5108-4cf1-add5-2fd389e4a889" />

**Comment by lihsai0** [comment]
> ok, the new helix amp jump feature also can use in vim mode, in keymap.json
> 
>  {
>     "context": "vim_mode == normal || vim_mode == visual",
>     "bindings": {
>       "s": "vim::HelixJumpToWord",
>     },
>   },


```
Zed 1.0.0 
Commit: 5ec84a926ef83865afb92d2a3d1ca3b419572cf9
Version: 1.0.0+stable.257.5ec84a926ef83865afb92d2a3d1ca3b419572cf9
```

And it is not working. 😅

<img width="426" height="216" alt="Image" src="https://github.com/user-attachments/assets/babb6658-bf9f-4a13-89c3-24ad8ab34e6f">

**Comment by aizigao** [comment]
> > ok, the new helix amp jump feature also can use in vim mode, in keymap.json
> > {
> > "context": "vim_mode == normal || vim_mode == visual",
> > "bindings": {
> > "s": "vim::HelixJumpToWord",
> > },
> > },
> 
> ```
> Zed 1.0.0 
> Commit: 5ec84a926ef83865afb92d2a3d1ca3b419572cf9
> Version: 1.0.0+stable.257.5ec84a926ef83865afb92d2a3d1ca3b419572cf9
> ```
> 
> And it is not working. 😅
> 
> <img alt="Image" width="426" height="216" src="https://private-user-images.githubusercontent.com/17806506/585776135-babb6658-bf9f-4a13-89c3-24ad8ab34e6f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Nzc1MTU4NTksIm5iZiI6MTc3NzUxNTU1OSwicGF0aCI6Ii8xNzgwNjUwNi81ODU3NzYxMzUtYmFiYjY2NTgtYmY5Zi00YTEzLTg5YzMtMjRhZDhhYjM0ZTZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA0MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNDMwVDAyMTkxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcyZjc3NWQ2MmJlNDQ0ODYyYjkwN2E5MjA5OGIxMzYwZWQyZTg2Yjc5NWYzNTg3NTU1ZDY4NTJiYzIzODRiMjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.puQfjQF_bT-5WVuUP4KiWifNEqpfzQBsbdPq_fjpfBI">

yes  I use Zed Preview 1.1.2

**Comment by GoldStrikeArch** [comment]
Okay, so I made a [proper hookup](https://github.com/zed-industries/zed/pull/55492) for the Vim mode to use a new "jump-to-word" (which was originally done for the Helix mode only, so it was not supposed to work in Vim mode in the first place 😅 ), feel free to checkout that branch and leave any comments if it doesn't work (the current keymap is `g z`, as `g w` is taken in Vim) as you expected (because again, it does reuse the current "Helix amp jump" algorithm and how we find the labels, etc). I never used vim/neovim so I don't know how it should behave but we can adjust it further if needed to tailor the specific "easymotions" or "flash" plugins (I suppose those a plugins?) implementation

**Comment by arcqiufeng** [comment]
> Okay, so I made a [proper hookup](https://github.com/zed-industries/zed/pull/55492) for the Vim mode to use a new "jump-to-word" (which was originally done for the Helix mode only, so it was not supposed to work in Vim mode in the first place 😅 ), feel free to checkout that branch and leave any comments if it doesn't work (the current keymap is `g z`, as `g w` is taken in Vim) as you expected (because again, it does reuse the current "Helix amp jump" algorithm and how we find the labels, etc). I never used vim/neovim so I don't know how it should behave but we can adjust it further if needed to tailor the specific "easymotions" or "flash" plugins (I suppose those a plugins?) implementation

Currently, this jump feature designed exclusively for Helix is already functional in Vim. However, for pure Vim users who do not use Helix, I believe there are two aspects that may feel uncomfortable:
1. In Vim mode, users generally expect to be in a non-selected state after jumping.
2. In Vim mode, users usually want the cursor to land at the start of a word rather than the end after jumping.

The above is just my personal opinion, and I am not a highly proficient Vim user. I welcome corrections from Vim experts.

**Comment by GoldStrikeArch** [comment]
@arcqiufeng 

> In Vim mode, users generally expect to be in a non-selected state after jumping.
> In Vim mode, users usually want the cursor to land at the start of a word rather than the end after jumping.

Yes, both of those things are adjusted in that PR

**Comment by arcqiufeng** [comment]
> [@arcqiufeng](https://github.com/arcqiufeng)
> 
> > In Vim mode, users generally expect to be in a non-selected state after jumping.
> > In Vim mode, users usually want the cursor to land at the start of a word rather than the end after jumping.
> 
> Yes, both of those things are adjusted in that PR

Thanks a lot for your work. The jump-to-word feature is way better and more efficient for CJK users than typing letters to search. For people working with plain CJK text, features like "sneak" or the native `f` key aren't really useful. Flash also has the jump-to-word feature—without it, it'd be pretty much useless for CJK text.

I also want to mention something really nice about flash.nvim: when it shows jump labels, the syntax highlighting of the rest of the content completely turns grayscale, so the labels really stand out. Right now, the jump labels just blend in with the theme's colors. If the theme isn't chosen well, the labels become totally unreadable.

These are just a few suggestions. Whether you take them or not, I want to thank you again for adding jump-to-word support.

**Comment by GoldStrikeArch** [comment]
@arcqiufeng 

> I also want to mention something really nice about flash.nvim: when it shows jump labels, the syntax highlighting of the rest of the content completely turns grayscale, so the labels really stand out. Right now, the jump labels just blend in with the theme's colors. If the theme isn't chosen well, the labels become totally unreadable.

You can override the jump label colors using this:

```json
{
  "theme_overrides": {
    "One Dark": {
      "vim.helix_jump_label.foreground": "#ffffff"
    }
  }
}

```

IMO it is not good if you turn off all syntax highlight or make the other parts "gray" as there would be a lot of flickering if you are doing multiple jumps in a row (the whole visible viewport would flicker as most people use syntax highlight). But if it is really easier to use to some people we can make this behavior as optional setting 🤔

**Comment by arcqiufeng** [comment]
Yeah, it's really just a matter of personal preference. I did see the config you mentioned, but for some high-contrast, colorful themes, the labels might still not stand out enough. As for the flickering issue, personally I don't really mind it.


I mostly use Vim mode in VS Code, and the flash and easymotion plugins there use the grayscale mode. The flash.nvim plugin I use in Neovim does the same. Personally, I find that mode more comfortable.


If you're open to providing an option for it, I'd really appreciate it. And I don't want my wishlist item to take up too much of your time either.


Thanks again for your work!




---Original---
From: "Mikhail ***@***.***&gt;
Date: Wed, May 13, 2026 14:10 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [zed-industries/zed] Flash.nvim style search in a document. (Issue #14801)


GoldStrikeArch left a comment (zed-industries/zed#14801)
 
@arcqiufeng
  
I also want to mention something really nice about flash.nvim: when it shows jump labels, the syntax highlighting of the rest of the content completely turns grayscale, so the labels really stand out. Right now, the jump labels just blend in with the theme's colors. If the theme isn't chosen well, the labels become totally unreadable.
  
You can override the jump label colors using this:
 {   "theme_overrides": {     "One Dark": {       "vim.helix_jump_label.foreground": "#ffffff"     }   } } 
 
IMO it is not good if you turn off all syntax highlight or make the other parts "gray" as there would be a lot of flickering if you are doing multiple jumps in a row (the whole visible viewport would flicker as most people use syntax highlight). But if it is really easier to use to some people we can make this behavior as optional setting 🤔
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by GoldStrikeArch** [comment]
@arcqiufeng @951mmm 

> If you're open to providing an option for it, I'd really appreciate it. And I don't want my wishlist item to take up too much of your time either.

Sure, I can do it. Can you please open a dedicated issue/discussion about this? I think it is not that hard to add this behavior

**Comment by arcqiufeng** [comment]
> [@arcqiufeng](https://github.com/arcqiufeng) [@951mmm](https://github.com/951mmm)
> 
> > If you're open to providing an option for it, I'd really appreciate it. And I don't want my wishlist item to take up too much of your time either.
> 
> Sure, I can do it. Can you please open a dedicated issue/discussion about this? I think it is not that hard to add this behavior

I started a discussion here: https://github.com/zed-industries/zed/discussions/56621

**Comment by Marrieto** [comment]
It's a shame that the search to jump feature isn't working yet, but I'm happy for the helix implementation that we can use now, it makes it so that I can actually start using Zed 👍

**Comment by Benjamin-van-Heerden** [comment]
What about flash? Don't mean to be ungrateful, helix jump is cool, but flash is, imo, better and more intuitive. It seems like all of the pieces are there to do flash with the generic navigation overlays. 

One example where I think flash is better is when you want to move to a bracket or to a dash inside a word. 

Has to be said that helix jump is a bit faster since you don't need to type out an intermediate subword before you are able to navigate so don't get me wrong if this is what I get I am already happy, but I like being able to place the cursor at exact positions.

**Comment by ccidral** [comment]
> What about flash? Don't mean to be ungrateful, helix jump is cool, but flash is, imo, better and more intuitive. It seems like all of the pieces are there to do flash with the generic navigation overlays.

I'm with you here. Maybe I'm missing something in this conversation but I don't understand why the discussion is about flash vs helix when it should be about adding core capabilities to zed that allow people to write extensions/plugins like flash-style, helix-style, or whatever style of search & jump. The specific style shouldn't be baked in zed, we just need editor support to implement those things as extensions.

**Comment by GoldStrikeArch** [comment]
> Maybe I'm missing something in this conversation but I don't understand why the discussion is about flash vs helix when it should be about adding core capabilities to zed that allow people to write extensions/plugins like flash-style, helix-style, or whatever style of search & jump.

This is unlikely to happen soon and currently the extensions API is limited to things like LSP/themes and it doesn't export that many things from the `Editor` itself. The thing is that it is not an easy task to define the boundaries what can/should be exported from the `Editor` (and it might require some refactoring there as well which is also a task on its own) and overall will require way more work than just implementing "flash jump" into `vim.rs` itself (similar to what I did with helix jump as the logic would be different and vim specific it would likely live somewhere near `vim.rs`).

Anyway, my point is that ATM it is simpler and faster to implement something like flash directly in the Zed's `vim` crate

**Comment by arcqiufeng** [comment]
Here’s where things stand: @GoldStrikeArch first implemented Helix’s native jump feature in Zed’s Helix mode, and then ported that same functionality to Zed’s Vim mode. That means the Vim mode now has part of what Flash.nvim offers — at least the ability to jump to any word.

Everyone already thought the hardest part was behind us, and now that the port is done, finishing the rest of Flash’s features should be just a matter of time.

The sad part is, Zed’s current plugin system can’t handle this kind of advanced functionality. We have to rely on core changes by skilled developers. So a big thanks to @GoldStrikeArch for getting the ball rolling — first by showing it in Helix mode, and then by adapting it to Vim mode. Now we’re really hoping to see full Flash (search + jump) implemented on top of this foundation.

**Comment by arcqiufeng** [comment]
Just a quick add as a CJK language user — in non‑code editing contexts, what they call "jump to word" really just jumps to the beginning of a sentence. That’s because our languages don’t use spaces to separate words. Nothing we can do about it, which is a bit of a bummer. To actually segment CJK text, you’d need to understand the meaning — simple algorithms just won’t cut it. Right now it basically requires an LLM. Even nvim’s flash.nvim can’t do that.

**Comment by GoldStrikeArch** [comment]
@arcqiufeng 

>  That’s because our languages don’t use spaces to separate words.

But that would mean that regular Vim/Helix motions also doesn't work there, right?

**Comment by cpbotha** [comment]
@GoldStrikeArch 
> [@arcqiufeng](https://github.com/arcqiufeng)
> 
> > That’s because our languages don’t use spaces to separate words.
> 
> But that would mean that regular Vim/Helix motions also doesn't work there, right?

I think you're right about "normal" vim/helix word motions, but with flash in neovim, you type the first two characters to get highlights. This works for mid-word targets as well, so it would work to jump anywhere to within CJK space-less sentences.

**Comment by arcqiufeng** [comment]
> But that would mean that regular Vim/Helix motions also doesn't work there, right?

Yes. Without any special handling for CJK word boundaries (e.g., word segmentation), w treats punctuation as word separators. For CJK text, that often means an entire sentence is treated as a single word. In Vim or Neovim, there are some workarounds — for example, jieba.vim is a Chinese word segmentation plugin. But overall, there is no lightweight solution.

> type the first two characters to get highlights

For Chinese, this is very inefficient because Chinese characters must be typed via an IME. So Flash is very useful when writing code, but when editing CJK documents, unless there is a good word segmentation tool, you end up jumping to the nearest match and then moving with hjkl.

**Comment by GoldStrikeArch** [comment]
Hi everyone,

I made the [PR](https://github.com/zed-industries/zed/pull/59188) to implement the basics of nvim.flash, can you please try it out and leave a feedback? I am not a Vim/neovim user myself, so I just watched some youtube videos and looked at the nvim.flash source code while implementing it. So it is highly likely that I missed some things 😅

**Comment by www159-used** [comment]
> Hi everyone,
> 
> I made the [PR](https://github.com/zed-industries/zed/pull/59188) to implement the basics of nvim.flash, can you please try it out and leave a feedback? I am not a Vim/neovim user myself, so I just watched some youtube videos and looked at the nvim.flash source code while implementing it. So it is highly likely that I missed some things 😅

i love u 🥰

**Comment by fang2hou** [comment]
> But that would mean that regular Vim/Helix motions also doesn't work there, right?
> Yes. Without any special handling for CJK word boundaries (e.g., word segmentation), w treats punctuation as word separators. For CJK text, that often means an entire sentence is treated as a single word. In Vim or Neovim, there are some workarounds — for example, jieba.vim is a Chinese word segmentation plugin. But overall, there is no lightweight solution.
>  type the first two characters to get highlights
> For Chinese, this is very inefficient because Chinese characters must be typed via an IME. So Flash is very useful when writing code, but when editing CJK documents, unless there is a good word segmentation tool, you end up jumping to the nearest match and then moving with hjkl.

You can try my new [flash-cjk.nvim](https://github.com/fang2hou/flash-cjk.nvim), and the core matcher is actually written in Rust.

**Comment by skuong** [comment]
I want this so bad. It's the biggest missing piece for me ever since I fully switched to Zed.
Zed team, please do something
